server/punishments.ts

```diff
@@ -108,87 +108,123 @@ class NestedPunishmentMap extends Map<RoomID, Map<string, Punishment>> {
 		}
 	}
 }
+
 /*********************************************************
- * Persistence
+ * Storage
  *********************************************************/
 
-export const Punishments = new class {
-	/**
-	 * ips is an ip:punishment Map
-	 */
-	ips = new PunishmentMap();
-	/**
-	 * userids is a userid:punishment Map
-	 */
-	userids = new PunishmentMap();
-	/**
-	 * roomUserids is a roomid:userid:punishment nested Map
-	 */
-	roomUserids = new NestedPunishmentMap();
-	/**
-	 * roomIps is a roomid:ip:punishment Map
-	 */
-	roomIps = new NestedPunishmentMap();
-	/**
-	 * sharedIps is an ip:note Map
-	 */
-	sharedIps = new Map<string, string>();
-	/**
-	 * Connection flood table. Separate table from IP bans.
-	 */
-	cfloods = new Set<string>();
-	/**
-	 * punishType is an allcaps string, for global punishments they can be
-	 * anything in the punishmentTypes map.
-	 *
-	 * This map can be extended with custom punishments by chat plugins.
-	 *
-	 * Keys in the map correspond to punishTypes, values signify the way
-	 * they should be displayed in /alt
-	 *
-	 */
-	punishmentTypes = new Map<string, string>([
-		['LOCK', 'locked'],
-		['BAN', 'globally banned'],
-		['NAMELOCK', 'namelocked'],
-	]);
-	/**
-	 * For room punishments, they can be anything in the roomPunishmentTypes map.
-	 *
-	 * This map can be extended with custom punishments by chat plugins.
-	 *
-	 * Keys in the map correspond to punishTypes, values signify the way they
-	 * should be displayed in /alt.
-	 * By default, this includes:
-	 * - 'ROOMBAN'
-	 * - 'BLACKLIST'
-	 * - 'BATTLEBAN'
-	 * - 'MUTE' (used by getRoomPunishments)
-	 *
-	 */
-	roomPunishmentTypes = new Map<string, string>([
-		['ROOMBAN', 'banned'],
-		['BLACKLIST', 'blacklisted'],
-		['BATTLEBAN', 'battlebanned'],
-		['MUTE', 'muted'],
-	]);
-	constructor() {
-		setImmediate(() => {
-			// tslint:disable-next-line: no-floating-promises
-			Punishments.loadPunishments();
-			// tslint:disable-next-line: no-floating-promises
-			Punishments.loadRoomPunishments();
-			// tslint:disable-next-line: no-floating-promises
-			Punishments.loadBanlist();
-			// tslint:disable-next-line: no-floating-promises
-			Punishments.loadSharedIps();
-		});
+class PunishmentsStorage {
+	static connect(type: string) {
+		let storage;
+		if (type === 'memory') {
+			storage = PunishmentsMemoryStorage;
+		} else if (type === 'tsv') {
+			storage = PunishmentsTsvStorage;
+		} else {
+			Monitor.log(`Unrecongized punishments database type: ${type}, defaulting to memory storage.`);
```

Comment:
I wanted to crash here, but it'd put the server under lockdown because it happens in the first thirty minutes of the server starting up, which isn't ideal, so I compromised with logging to `stdout` and the Staff room.